internal/controller/api_to_config.go

```diff
@@ -73,98 +73,178 @@ func neighborToFRR(n v1beta1.Neighbor, ipv4Prefixes, ipv6Prefixes []string) (*fr
 		Addr: n.Address,
 		Port: n.Port,
 		// Password:       n.Password, TODO password as secret
-		Advertisements: make([]*frr.AdvertisementConfig, 0),
-		IPFamily:       neighborFamily,
-		EBGPMultiHop:   n.EBGPMultiHop,
+		IPFamily:     neighborFamily,
+		EBGPMultiHop: n.EBGPMultiHop,
 	}
 
-	advs := map[string]*frr.AdvertisementConfig{}
-	if n.ToAdvertise.Allowed.Mode == v1beta1.AllowAll {
+	res.Outgoing, err = toAdvertiseToFRR(n.ToAdvertise, ipv4Prefixes, ipv6Prefixes)
+	if err != nil {
+		return nil, err
+	}
+	res.Incoming = toReceiveToFRR(n.ToReceive)
+	return res, nil
+}
+
+func toAdvertiseToFRR(toAdvertise v1beta1.Advertise, ipv4Prefixes, ipv6Prefixes []string) (frr.AllowedOut, error) {
+	advsV4, advsV6 := prefixesToMap(toAdvertise, ipv4Prefixes, ipv6Prefixes)
+	communities, err := communityPrefixesToMap(toAdvertise.PrefixesWithCommunity)
+	if err != nil {
+		return frr.AllowedOut{}, err
+	}
+	err = setCommunitiesToAdvertisements(advsV4, communities, ipfamily.IPv4)
+	if err != nil {
+		return frr.AllowedOut{}, err
+	}
+	err = setCommunitiesToAdvertisements(advsV6, communities, ipfamily.IPv6)
+	if err != nil {
+		return frr.AllowedOut{}, err
+	}
+	res := frr.AllowedOut{
+		PrefixesV4: sortMap(advsV4),
+		PrefixesV6: sortMap(advsV6),
+	}
+	return res, nil
+}
+func prefixesToMap(toAdvertise v1beta1.Advertise, ipv4Prefixes, ipv6Prefixes []string) (map[string]*frr.OutgoingFilter, map[string]*frr.OutgoingFilter) {
+	resV4 := map[string]*frr.OutgoingFilter{}
+	resV6 := map[string]*frr.OutgoingFilter{}
+	if toAdvertise.Allowed.Mode == v1beta1.AllowAll {
 		for _, p := range ipv4Prefixes {
-			advs[p] = &frr.AdvertisementConfig{Prefix: p, IPFamily: ipfamily.IPv4}
-			res.HasV4Advertisements = true
+			resV4[p] = &frr.OutgoingFilter{Prefix: p, IPFamily: ipfamily.IPv4}
 		}
 		for _, p := range ipv6Prefixes {
-			advs[p] = &frr.AdvertisementConfig{Prefix: p, IPFamily: ipfamily.IPv6}
-			res.HasV6Advertisements = true
+			resV6[p] = &frr.OutgoingFilter{Prefix: p, IPFamily: ipfamily.IPv6}
 		}
+		return resV4, resV6
 	}
-
-	for _, p := range n.ToAdvertise.Allowed.Prefixes {
+	for _, p := range toAdvertise.Allowed.Prefixes {
 		family := ipfamily.ForCIDRString(p)
 		switch family {
 		case ipfamily.IPv4:
-			res.HasV4Advertisements = true
+			resV4[p] = &frr.OutgoingFilter{Prefix: p, IPFamily: family}
 		case ipfamily.IPv6:
-			res.HasV6Advertisements = true
+			resV6[p] = &frr.OutgoingFilter{Prefix: p, IPFamily: family}
 		}
-
-		// TODO: check that the prefix matches the passed IPv4/IPv6 prefixes
-		advs[p] = &frr.AdvertisementConfig{Prefix: p, IPFamily: family}
```

Comment:
Nope, I missed it. I will leave a TODO because doing it while we process seems a bit ugly, so it might worth having an ad hoc validate function. Let's wait after we do the merging thing and the dust settles